// typhonjs-escomplex-commons/dist/module/report/control/ModuleScopeControl.js
'use strict';

Object.defineProperty(exports, "__esModule", {
   value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _ClassReport = require('../ClassReport');

var _ClassReport2 = _interopRequireDefault(_ClassReport);

var _ClassMethodReport = require('../ClassMethodReport');

var _ClassMethodReport2 = _interopRequireDefault(_ClassMethodReport);

var _ModuleMethodReport = require('../ModuleMethodReport');

var _ModuleMethodReport2 = _interopRequireDefault(_ModuleMethodReport);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ModuleScopeControl = function () {
   function ModuleScopeControl(moduleReport) {
      _classCallCheck(this, ModuleScopeControl);

      this._report = moduleReport;

      /**
       * Stores the current class report scope stack which is lazily created in `createScope`.
       * @type {Array<ClassReport>}
       */
      this._scopeStackClass = [];

      /**
       * Stores the current method report scope stack which is lazily created in `createScope`.
       * @type {Array<ClassMethodReport|ModuleMethodReport>}
       */
      this._scopeStackMethod = [];
   }

   /**
    * Creates a report scope when a class or method is entered.
    *
    * @param {object}   newScope - An object hash defining the new scope including:
    * ```
    * (string) type - Type of report to create.
    * (string) name - Name of the class or method.
    * (number) lineStart - Start line of method.
    * (number) lineEnd - End line of method.
    * (number) paramCount - (For method scopes) Number of parameters for method.
    * ```
    *
    * @return {object}
    */


   _createClass(ModuleScopeControl, [{
      key: 'createScope',
      value: function createScope() {
         var newScope = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

         var report = void 0;

         if ((typeof newScope === 'undefined' ? 'undefined' : _typeof(newScope)) !== 'object') {
            throw new TypeError('createScope error: \'newScope\' is not an \'object\'.');
         }

         if (typeof newScope.type !== 'string') {
            throw new TypeError('createScope error: \'newScope.type\' is not a \'string\'.');
         }

         if (typeof newScope.name !== 'string') {
            throw new TypeError('createScope error: \'newScope.name\' is not a \'string\'.');
         }

         if (!Number.isInteger(newScope.lineStart)) {
            throw new TypeError('createScope error: \'newScope.lineStart\' is not an \'integer\'.');
         }

         if (!Number.isInteger(newScope.lineEnd)) {
            throw new TypeError('createScope error: \'newScope.lineEnd\' is not an \'integer\'.');
         }

         switch (newScope.type) {
            case 'class':
               report = new _ClassReport2.default(newScope.name, newScope.lineStart, newScope.lineEnd);
               this._report.classes.push(report);
               this._scopeStackClass.push(report);
               break;

            case 'method':
               {
                  if (!Number.isInteger(newScope.paramCount)) {
                     throw new TypeError('createScope error: \'newScope.paramCount\' is not an \'integer\'.');
                  }

                  // If an existing class report / scope exists also push the method to the class report.
                  var classReport = this.getCurrentClassReport();

                  if (classReport) {
                     report = new _ClassMethodReport2.default(newScope.name, newScope.lineStart, newScope.lineEnd, newScope.paramCount);
                     classReport.methods.push(report);
                  } else {
                     report = new _ModuleMethodReport2.default(newScope.name, newScope.lineStart, newScope.lineEnd, newScope.paramCount);

                     // Add this report to the module methods as there is no current class report.
                     this._report.methods.push(report);
                  }

                  this._scopeStackMethod.push(report);

                  break;
               }

            default:
               throw new Error('createScope error: Unknown scope type (' + newScope.type + ').');
         }

         return report;
      }

      /**
       * Returns the current class report.
       *
       * @returns {ClassReport}
       */

   }, {
      key: 'getCurrentClassReport',
      value: function getCurrentClassReport() {
         if (!Array.isArray(this._scopeStackClass)) {
            return void 0;
         }
         return this._scopeStackClass.length > 0 ? this._scopeStackClass[this._scopeStackClass.length - 1] : void 0;
      }

      /**
       * Returns the current method report.
       *
       * @returns {ClassMethodReport|ModuleMethodReport}
       */

   }, {
      key: 'getCurrentMethodReport',
      value: function getCurrentMethodReport() {
         if (!Array.isArray(this._scopeStackMethod)) {
            return void 0;
         }
         return this._scopeStackMethod.length > 0 ? this._scopeStackMethod[this._scopeStackMethod.length - 1] : void 0;
      }

      /**
       * Pops a report scope.
       *
       * @param {object}   scope - An object hash defining the scope including:
       * ```
       * (string) type - Type of report scope to pop off the stack.
       * ```
       */

   }, {
      key: 'popScope',
      value: function popScope(scope) {
         if ((typeof scope === 'undefined' ? 'undefined' : _typeof(scope)) !== 'object') {
            throw new TypeError('popScope error: \'scope\' is not an \'object\'.');
         }

         if (typeof scope.type !== 'string') {
            throw new TypeError('popScope error: \'scope.type\' is not a \'string\'.');
         }

         switch (scope.type) {
            case 'class':
               this._scopeStackClass.pop();
               break;

            case 'method':
               this._scopeStackMethod.pop();
               break;

            default:
               throw new Error('popScope error: Unknown scope type (' + scope.type + ').');
         }
      }
   }]);

   return ModuleScopeControl;
}();

exports.default = ModuleScopeControl;
module.exports = exports['default'];
